seps/sep-0002.md

```diff
@@ -0,0 +1,102 @@
+# Use-variants: for packages with conflicting libraries
+
+## The problem
+
+The `intel-oneapi-mkl` package is installed with many libraries. Some
+of these libraries define the same symbols but differ in their
+implementation. For example there is a set of libraries built with
+threading support, and another set which offers the same
+functionality but are built without threading support.
+
+Right now, the only mechanism Spack offers to choose between these
+libraries is a variant: you could add a `threading` variant and then
+choose which libraries to supply to dependents based on the value
+of that variant. The problem with a variant is that different
+variant values generate distinct installations of a package: we would
+install `intel-oneapi-mkl` once for each choice of `threading`, even
+though the installations would be the same.
+
+See also: https://github.com/spack/spack/discussions/22749
+
+## Proposed Changes
+
+* Modify variant declarations with a flag to indicate that the variant
+  determines how the package is used rather than how it is installed
+* During concretization, this variant can be treated like any other
+  variant.
+* After concretization, the variant is stripped from the node and
+  stored on the edge connecting the node to its parent.
+
+## Example Use Cases
+
+### Intel oneAPI MKL
+
+Described in the introduction. This would be resolved by adding a
+`threading` variant:
+
+```
+variant('threading', values=['openmp', 'sequential'],
+        interface=True)
```

Comment:
I think we should have a more generic argument that can take a single value or a list of them:
```suggestion
variant('threading', values=['openmp', 'sequential'], type='interface')
```
In this way we would use a single directive to declare variants that are relevant for both deploying a package and using it. With the proposed interface such a case would need two declarations:
```
variant('libs', values=('shared', 'static'), multi=True, interface=True)
variant('libs', values=('shared', 'static'), multi=True, interface=False)
```
while using a more generic argument one suffices:
```
variant('libs', values=('shared', 'static'), multi=True, type=('deploy', 'interface'))
```
Another minor benefit of this would be that we'll leave the type of variants open for extensions, even though it's very likely that with these two different types ("deploy" and "interface") we'll be good for a long time.

Reply:
I agree with this though not necessarily on the names "interface" and "deploy", but we can work those out later.

Reply:
@tgamblin Sure, even "type" as the name of the argument is just a placeholder at the moment.